### PR TITLE
Classic Theme Helper: Require Custom Content Types (portfolios) from the package

### PR DESCRIPTION
**Pull request by coder-karen**
Fixes https://github.com/Automattic/vulcan/issues/476
Fixes https://github.com/Automattic/vulcan/issues/478
Fixes https://github.com/Automattic/vulcan/issues/477

Note: Jetpack will need to be updated on WoA (ideally) before Jetpack Mu WPcom features to prevent a small issue where two toggle attempts are needed (so this PR would ideally be merged first thing on Monday morning EU time).

## Proposed changes:

* This PR ensures we are requiring the newly moved Custom Content Types code (and specifically with portfolio related files) from the Classic Theme Helper package, in the Jetpack, Jetpack Mu WPcom plugins and Classic Theme Helper plugin.
* This PR also adds in deprecations to related functions in the `custom-content-types.php` and the `portfolios.php` module files, removing several private functions in `portfolios.php` and adding custom calling functions as well.
* A larger change is in `class-portfolios.php`, related to function load time:
  * Instead of directly calling the `init` function from the `Jetpack_Portfolio` class which looked like it would need to happen so that the toggle state (from the Jetpack settings page) would correctly be applied and used in the `Jetpack_Portfolio` class, I added logic in the classes `__construct` function to only directly call the `maybe_register_cpt` function if the portfolio option exists, and if it doesn't it makes sure the call to that function is hooked into init.
  * The main reason I avoided using init directly is that we keep initialization and lifecycle management outside of the package and within the plugin (meaning that Jetpack controls when and how the packages classes or methods are initialized, not the package initiliazing itself).
* The Jetpack > Settings > Writing page has also had changes:
  * ~If `this.props.getOptionValue` is undefined for `jetpack_portfolio`, we check `this.props.settings.jetpack_portfolio`.~
  * ~`componentDidMount` runs once, when the component is added to the DOM, and ensures that we're setting the portfolio state to the correct value. Without it, the toggle doesn't work as expected (particularly when toggling off, which usually requires two attempts).~
  * Testimonials state now also has a default of false if the value is undefined, as when it is undefined this also messes with the portfolio toggle functionality. Testimonials functionality will be updated when testimonials are moved to the Classic Theme Helper package, which will come later.
  * The changes in the `ToggleControl` when rendering are to make sure we don't get a console warning related to 'A component is changing a controlled input to be uncontrolled'.
* This also fixes what appears to be a bug with the original implementation given the existing code - on theme activation for a theme that supports the Portfolio CPT, this will now enable the feature (even if none exist).

	

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

https://github.com/Automattic/vulcan/issues/476
p1725985286376189-slack-C05PV073SG3

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

**Self-hosted:**

* Ensure you can activate Portfolios via Jetpack > Settings > Writing at `wp-admin/admin.php?page=jetpack#/writing` (see the Custom Content Types section).
* Test that Portfolio functionality works as expected: Create a new Portfolio via the Portfolio wp-admin menu. View it. Add it to another post via the portfolio shortcode (`[portfolio]`).
* Test that a theme that supports Jetpack Portfolios auto-activates it: Make sure Portfolios are not active, and then install and activate the Sketch theme. Portfolios should now be active. If you've already tested this, to retest you may need to toggle the Portfolios option on and off a few times for it to work again when reactivating Sketch (same in trunk).
* Try visiting `/wp-admin/edit.php?post_type=jetpack-portfolio` when Portfolio is toggled off - you should see an 'invalid post type' message, and on the front-end on a post which made use of the Portfolio shortcode, you should see the unrendered shortcode. 
* Generally test that toggling on and off the Portfolios feature results in expected Portfolio functionality being visible and not visible - including the Portfolio menu item display in the main wp-admin menu.
* Test with Testimonials active and not active as well, in terms of how that impacts Portfolios, though there are some testing quirks around that depending on the theme (see below).
* Make sure there are no deprecation notices in error logs while testing.

**WoA:**

* As with self-hosted, however if only the WordPress.com Features branch is active then Portfolio will not auto-toggle on when activating a theme that supports CPTs (Jetpack will need to be updated on WoA first).

**Simple:**

This will need to be tested with this diff applied to prevent fatal errors due to duplicated function names: D161146-code

* Apply the branches as per the generated comment below. Testing instructions roughly follow self-hosted but with several differences:
  * Doing similar testing as with self-hosted, the difference is that instead of checking the Jetpack > Settings > Writing page, you check the Settings > Writing page at `/wp-admin/options-writing.php`. Toggling off isn't possible when a theme supports Portfolios or Testimonials. But toggling off with another theme like Dara, then activating Sketch and refreshing the `options-writing.php` page, you should see Portfolio is auto-enabled due to the theme, and the Portfolio menu item exists.
  * I'd add that I was unable to fully test the theme activation flow though as activating Sketch required me to unsandbox WPcom, then resandbox when it was activated.
  * Note as well that with just the Jetpack branch applied, the Portfolio CPT option (if active) gets disabled upon activating the branch. However this shouldn't be a problem as both will get updated at the same time.

**Classic Theme Helper plugin:**

* If testing locally, with the PR applied make sure you have installed and built the Classic Helper Theme plugin: `jetpack install plugins/classic-theme-helper-plugin`, `jetpack build plugins/classic-theme-helper-plugin`, then activate the plugin.
* Also if testing locally, remove `"automattic/jetpack-classic-theme-helper": "@dev"`, from the Jetpack plugin `composer.json` file. Then run `tools/fixup-project-versions.sh`, and then `jetpack install plugins/jetpack` and` jetpack build plugins/jetpack`.
* If testing using the Beta tester plugin, activate the branch for the Classic Theme Helper plugin.
* If testing using the Beta tester plugin, delete the `jetpack-classic-theme-helper` package folder from `jetpack-dev/jetpack-vendor/automattic`.
* Follow the steps under 'Self-hosted' to test Portfolio CPTs.

<details>
<summary>Some existing testing quirks:</summary>
- It is expected when changing theme that if the theme doesn't support the CPT (Testimonials or Portfolios), those will individually be toggled off if there are none of them (no testimonials, no portfolio items). Also if you enable Testimonials on Sketch, it will enable Portfolios as well. Also in Dara, toggling Portfolios on toggle Testimonials on.<br/>
- In trunk, activating a theme with supports the Portfolio CPT will only toggle on Portfolios if there is an existing Portfolio item - unless you had Testimonials active as well, in which case Portfolios will activate. This is the only behaviour that differs from this branch, which will toggle on Portfolio regardless, as that appears to be the expected functionality.

</details>

### CODE REVIEW

File: projects/packages/classic-theme-helper/src/custom-post-types/class-jetpack-portfolio.php [hunk 1]
```diff
@@ -59,7 +59,14 @@ public function __construct() {
...
 
-			$this->maybe_register_cpt();
+			// If portfolio cpt is enabled, hook into init to register the CPT, otherwise run maybe_register_cpt immediately to deregister.
+			if ( get_option( self::OPTION_NAME, '0' ) ) {
+				$this->maybe_register_cpt();
+
+			} else {
+				add_action( 'init', array( $this, 'maybe_register_cpt' ) );
+
+			}
```
Comment (coder-karen) [inline projects/packages/classic-theme-helper/src/custom-post-types/class-jetpack-portfolio.php]: To explain the extra logic in `__construct`: We immediately run `maybe_register_cpt` if the portfolio option exists as it needs to be registered immediately. If it doesn't exist, we hook the function call into init to ensure it's only cleaned up after WP has fully loaded. 

In short this is to ensure `maybe_register_cpt` is run at the correct lifecycle stage when the portfolio option exists vs when it doesn't, as this was a noticeable issue when testing the feature via the package.

**Comment by coder-karen** [comment]
Modified `custom-content-types.jsx` thanks to some feedback from @sergeymitr in Slack: p1726097992511299/1725985286.376189-slack-C05PV073SG3

**Review by fgiannar — APPROVED** [review]
I tested all the different scenarios following the test instructions on self-hosted, WoA and Simple.
For Simple sites, I was also able to test the theme activation flow successfully. I had portfolio disabled, then activated Sketch and confirmed it was auto-enabled after the theme activation.
Great work here, Karen 👍 🚢

**Comment by coder-karen** [comment]
Changing label to 'in progress' temporarily to prevent any accidental merges until the timing is right to merge this.

**Pull request merged**